### PR TITLE
Use custom derive ser/de instead of std::io

### LINKED ISSUE #443: Add `no_std` canonical serialization and deserialization
The initial PR: https://github.com/FuelLabs/fuel-tx/pull/190

With our own canonical serialization/deserialization we can avoid the usage of `io::Read`/`io::Write` and be able to compile with `no_std` mode.

Also, it can remove code duplication and simplify the support of serialization for types.

### PR DESCRIPTION
**Pull request by Dentosal**
Closes #443. Initial work migrated from https://github.com/FuelLabs/fuel-tx/pull/190, and the improved and adapted to the current system.

This PR replaces hand-written type serialization code with a macro-derived ones. It removes a lot of error-prone boilerplate code and makes it easier to add new types. The old implementation using `io::{Read, Write}` was not `no_std` compatible, but the new generated code is. The new implmentation also allows computing many sizes at compile time.

The new serialization framework is called `canonical`. It's used similarly to `serde`:

```rust
use fuel_types::canonical::{Serialize, Deserialize};

#[derive(Serialize, Deserialize)]
pub struct ExampleStruct {
    inner: InnerStruct,
    dynamically_sized: Vec<u8>,
    primitive: u64,
    #[canonical(ignore)]
    primitive: u64,
}
```

Several datatype-level attributes are also accepted:, expecially for enums.

### structs

https://github.com/FuelLabs/fuel-vm/blob/4af7c6312561576498acdca2f62c2235c94ce2bd/fuel-derive/src/attribute.rs#L59-L61

### enums

https://github.com/FuelLabs/fuel-vm/blob/4af7c6312561576498acdca2f62c2235c94ce2bd/fuel-derive/src/attribute.rs#L81-L95

### CODE REVIEW

File: fuel-asm/src/lib.rs
Comment (Voxelot) [inline fuel-asm/src/lib.rs:9]: Can we re-enable this now that the impl work is done?

File: fuel-derive/Cargo.toml
Comment (Voxelot) [inline fuel-derive/Cargo.toml:3]:
```suggestion
version = { workspace = true }
```

File: fuel-derive/Cargo.toml
Comment (Voxelot) [inline fuel-derive/Cargo.toml:10]: version = { workspace = true }
authors = { workspace = true }
categories = ["cryptography::cryptocurrencies", "data-structures"]
edition = { workspace = true }
homepage = { workspace = true }
keywords = ["blockchain", "cryptocurrencies", "fuel-vm", "vm"]
license = "Apache-2.0"
repository = { workspace = true }

File: fuel-derive/Cargo.toml
Comment (Voxelot) [inline fuel-derive/Cargo.toml:11]:
```suggestion
description = "FuelVM (de)serialization derive macros for `fuel-vm` data structures."
```

File: fuel-derive/Cargo.toml
Comment (Voxelot) [inline fuel-derive/Cargo.toml:20]: newline
```suggestion
synstructure = "0.13"

```

File: fuel-tx/Cargo.toml
Comment (Voxelot) [inline fuel-tx/Cargo.toml:17]: This should be a workspace dep
```suggestion
fuel-derive = { workspace = true }
```

File: fuel-tx/src/lib.rs
Comment (Voxelot) [inline fuel-tx/src/lib.rs:8]: can we re-enable this?

File: fuel-tx/src/transaction/types/create.rs
Comment (Voxelot) [inline fuel-tx/src/transaction/types/create.rs:198]: There are some areas in the code below which are affected by this change:
image

File: fuel-types/Cargo.toml
Comment (Voxelot) [inline fuel-types/Cargo.toml:14]: Should use a workspace dep

File: fuel-types/src/canonical.rs
Comment (Voxelot) [inline fuel-types/src/canonical.rs:382]: why do we take the first byte and append it at the end?

File: fuel-types/src/lib.rs
Comment (Voxelot) [inline fuel-types/src/lib.rs:14]: isn't alloc always enabled by std? Seems like there should be a way to simplify this pattern.

File: fuel-derive/src/serialize.rs
Comment (Voxelot) [inline fuel-derive/src/serialize.rs:488]: Should this be removed?

File: fuel-types/src/canonical.rs
Comment (Voxelot) [inline fuel-types/src/canonical.rs:347]: Do we have any safety checks to ensure we aren't transmuting heap pointers here like `Box` etc?

File: fuel-types/src/canonical.rs
Comment (Voxelot) [inline fuel-types/src/canonical.rs:508]: How does this compare to the previous approach of using the memlayout + some manual logic for dynamic sizes? Are we incurring any new overhead by first traversing all serializable fields, converting each one into a slice, and then storing each of their lengths here?

Comment (Voxelot) [reply fuel-types/src/canonical.rs:508]: If the process of converting a field into a `&[u8]` has overhead, then we end up paying this twice for every call to `to_bytes()` since we get the size first to initialize a vec with enough capacity.

Comment (Voxelot) [reply fuel-types/src/canonical.rs:508]: Maybe we should consider calculating the static size of each type at compile time?

File: fuel-vm/src/constraints.rs
Comment (xgreenx) [inline fuel-vm/src/constraints.rs:67]:
```suggestion
        T: Deserialize,
```

File: fuel-vm/src/call.rs
Comment (xgreenx) [inline fuel-vm/src/call.rs:24]:
```suggestion
#[derive(Deserialize, Serialize)]
```

File: fuel-vm/src/call.rs
Comment (xgreenx) [inline fuel-vm/src/call.rs:72]:
```suggestion
    Deserialize,
    Serialize,
```

File: fuel-types/src/lib.rs
Comment (xgreenx) [reply fuel-types/src/lib.rs:14]: The main idea is that `canonical` s/d should be always available, we don't need to use `std` or `alloc` features.

File: fuel-types/src/canonical.rs
Comment (xgreenx) [inline fuel-types/src/canonical.rs:237]: You don't need to implement it manually. You can derive it because these types are defined in the same crate.

During the previous implementation, I defined traits in the `fuel-tx` which is why I should implemented them manually.

File: fuel-types/src/canonical.rs
Comment (xgreenx) [reply fuel-types/src/canonical.rs:382]: We iterate over all bytes(we use `capacity`, not length to do that because we have an empty vector with a pre-defined capacity) and add them to the current vector. Everything is expected=)

File: fuel-types/src/canonical.rs
Comment (xgreenx) [reply fuel-types/src/canonical.rs:347]: The `Box` works with raw pointers, so the memory layout for any `T` is the same. But maybe it is better to work directly with `&[u8]` to avoid any question in the future=)

File: fuel-types/src/canonical.rs
Comment (xgreenx) [reply fuel-types/src/canonical.rs:508]: It is the safest way to know the final size of the object. But yeah, we can optimize it later by introducing `STATIC_PART_SIZE`.

I'm okay with this during the first iteration of this transition.

Comment (Dentosal) [reply fuel-types/src/canonical.rs:508]: Static size *is* computed into `SIZE_STATIC` constant. The transition away from `SizeCalculator´ shoulnd't be too hard. I left that for a follow-up PR, but it might make sense to do it right now.

File: fuel-types/src/lib.rs
Comment (Voxelot) [reply fuel-types/src/lib.rs:14]: I think alloc might be required since the serialization uses vec

File: fuel-types/src/canonical.rs
Comment (Dentosal) [reply fuel-types/src/canonical.rs]: Migrated to derived impl in 34e6a048

File: fuel-tx/src/transaction/consensus_parameters/gas.rs
Comment (Dentosal) [reply fuel-tx/src/transaction/consensus_parameters/gas.rs:5]: Yep. Done in df32bbb8. The CI isn't catching this, but I'm unsure whether adding even more feature/crate combinations is a good idea regarding speed.

File: fuel-types/src/canonical.rs
Comment (Dentosal) [reply fuel-types/src/canonical.rs:295]: `T::UNALIGNED_BYTES` is only set for `u8` as the comment says. Types are also serialized bottom-up so this shouldn't be too much of an issue anyway. I'm still considering removing the `unsafe` usage, since I'm not sure that the performance impact will be meaningful.

File: fuel-derive/src/serialize.rs
Comment (Dentosal) [reply fuel-derive/src/serialize.rs:499]: Maybe? That's the main way to debug derive macros, so maybe I'll rather just clean it up a bit and document what it does instead.

File: fuel-types/src/lib.rs
Comment (Dentosal) [reply fuel-types/src/lib.rs]: I did more fine-grained gating, so it should be fixed now. Currently `alloc` is enabled by `std`, so maybe that could be simplified even more.

File: fuel-vm/src/constraints.rs
Comment (Dentosal) [reply fuel-vm/src/constraints.rs]: Done in df32bbb8

File: fuel-vm/src/call.rs
Comment (Dentosal) [reply fuel-vm/src/call.rs:73]: Done in f329eacf

File: fuel-vm/src/call.rs
Comment (Dentosal) [reply fuel-vm/src/call.rs]: Done in f329eacf

File: fuel-types/src/canonical.rs
Comment (Dentosal) [reply fuel-types/src/canonical.rs:331]: Yep, this is indeed correct

File: fuel-derive/src/attribute.rs
Comment (Dentosal) [reply fuel-derive/src/attribute.rs:18]: I totally agree here. Sadly Rust doesn't have reflection, and manually writing code for all possible fields is not something I like. Maybe just a list of allowed attributes plus duplicate checking would be good enough.

File: fuel-asm/Cargo.toml
Comment (Dentosal) [reply fuel-asm/Cargo.toml:31]: Not needed anymore, I think. I'll remove it.

File: CHANGELOG.md
Comment (Dentosal) [reply CHANGELOG.md:11]: True.

File: fuel-tx/src/transaction/repr.rs
Comment (Dentosal) [reply fuel-tx/src/transaction/repr.rs:8]: I've never had any issues, and it would be interesting to know what issues we had.

File: CHANGELOG.md
Comment (Dentosal) [reply CHANGELOG.md]: Fixed in 912a2786

File: fuel-types/src/canonical.rs
Comment (Dentosal) [reply fuel-types/src/canonical.rs]: It did make sense to do right now, as it required so much changes.

File: fuel-tx/src/transaction/types.rs
Comment (Dentosal) [reply fuel-tx/src/transaction/types.rs]: Done in 5bbdb451

File: fuel-tx/src/transaction/types.rs
Comment (Dentosal) [reply fuel-tx/src/transaction/types.rs]: Done in 5bbdb451

File: fuel-tx/src/transaction/consensus_parameters/gas.rs
Comment (Dentosal) [reply fuel-tx/src/transaction/consensus_parameters/gas.rs]: Done in 5bbdb451

File: fuel-asm/Cargo.toml
Comment (Dentosal) [reply fuel-asm/Cargo.toml]: Removed in a24f9a7a1742d3043b446250034447343b92e9c9

File: fuel-derive/src/deserialize.rs
Comment (Dentosal) [reply fuel-derive/src/deserialize.rs]: Done in 635dc638cc362ac52e79dcd7252f2e543fee9e62

File: fuel-derive/src/attribute.rs
Comment (Dentosal) [reply fuel-derive/src/attribute.rs]: Done in cf1870d4

File: fuel-derive/src/deserialize.rs
Comment (bvrooman) [inline fuel-derive/src/deserialize.rs:207]: Does this automatically support tuple structs as well?

File: fuel-derive/src/serialize.rs
Comment (bvrooman) [inline fuel-derive/src/serialize.rs:237]: It looks like every unsigned integer type is going to be serialized to take up 8 bytes (except 16 for `u128`) - is that correct? Is this part of the specification? 
Maybe using constants here, with a mention of the spec, would make this a little more clear and a little more flexible in case the specification ever changes.

File: fuel-derive/src/utils.rs
Comment (bvrooman) [inline fuel-derive/src/utils.rs:20]: Nice use of the new `OnceLock` feature

**Review by bvrooman — COMMENTED** [review]
I looked at about 35 files. I had some questions/minor suggestions for your discretion. I can keep reviewing later, but don't wait for me if you've already gotten a full review from the others.

File: fuel-derive/src/deserialize.rs
Comment (Dentosal) [reply fuel-derive/src/deserialize.rs:207]: Yes. See syn documentation here: https://docs.rs/syn/latest/syn/enum.Data.html

File: fuel-derive/src/attribute.rs
Comment (Dentosal) [reply fuel-derive/src/attribute.rs:68]: See field documentation for prefix:

https://github.com/FuelLabs/fuel-vm/blob/e048864f1a4b41e41b2f66f494614a6794e04bc0/fuel-derive/src/attribute.rs#L59-L61

So in practice it's used something like this:

```rust
#[derive(Serialize, Deserialize)]
#[canonical(prefix = 3u64)]
pub struct Prefixed3 {
  a: u64,
  b: u64,
}
```

Now `Prefixed3 { a: 5, b: 7 }` serializes to:

```
00 00 00 00 00 00 00 03 // prefix
00 00 00 00 00 00 00 05 // a
00 00 00 00 00 00 00 07 // b
```

This is especially useful with `#[canonical(inner_discriminant = PrefixType)]`:

```rust
#[derive(Serialize, Deserialize, TryFromPrimitive, IntoPrimitive)]
pub enum Discriminant {
  P1 = 1,
  P2 = 2,
}

#[derive(Serialize, Deserialize)]
#[canonical(prefix = Discriminant::P1)]
pub struct Prefixed1 {
  a: u64,
  b: u64,
}

#[derive(Serialize, Deserialize)]
#[canonical(prefix = Discriminant::P2)]
pub struct Prefixed2 {
  a: u64,
  data: Vec<u8>,
}


#[derive(Serialize, Deserialize]
#[canonical(inner_discriminant = Discriminant)]
pub enum Example {
  P1(Prefixed1),
  P2(Prefixed2),
}
```

Now when serializing `Example`, the discriminant isn't written to output, and instead the inner structs `Prefixed{1,2}` perform the write on serialization. When deserializing, the prefix is read to determine which variant to deserialize, but not consumed by the enum. The variant itself then re-reads the prefix.

This makes `Example` to not take more space than any of it's variants, when it can identify the variant by it's prefix instead. This is similar tricks that can be done with i.e. `union` in `C`. For our purposes this is useful since some structures must always start with a given identifier, and we would still like to warp over them using `enum`. I'm not sure if there's actually any reason those types are prefixed with their identifier, but I'd rather just replicate the old behavior exactly in this PR, and consider any changes in serialization behavior in follow-up PRs.

Comment (bvrooman) [reply fuel-derive/src/attribute.rs:68]: Great, thanks for the clear explanation. Is the prefix variable length or is it required to be a certain length, i.e. byte?

Comment (Dentosal) [reply fuel-derive/src/attribute.rs:68]: It's exactly one word, yes. It could be changed, but since that's the only use case now, I decided to keep it simple.

**Review by bvrooman — COMMENTED** [review]
Can we now remove the `mem_layout` macro and related code in this PR?

File: fuel-tx/src/transaction/types/input/coin.rs
Comment (bvrooman) [inline fuel-tx/src/transaction/types/input/coin.rs:180]: Can this line be replaced with `..Default::default()` for consistency with your changes below?

File: fuel-tx/src/transaction/types/input/message.rs
Comment (bvrooman) [inline fuel-tx/src/transaction/types/input/message.rs:221]: Can these be replaced with `..Default::default()`?

**Comment by bvrooman** [comment]
I've reviewed 66 files at this point - I will continue with the last stretch tomorrow.

File: fuel-derive/src/attribute.rs
Comment (Voxelot) [reply fuel-derive/src/attribute.rs]: Maybe a crate like [darling](https://crates.io/crates/darling) would be helpful?

File: fuel-derive/Cargo.toml
Comment (Voxelot) [inline fuel-derive/Cargo.toml:1]: We should also have a README.md for this crate that include status badges for ci and crates.io

File: fuel-derive/manual_impl_tx_serde.rs
Comment (Voxelot) [inline fuel-derive/manual_impl_tx_serde.rs:1]: How does this file get used? I can't find it being imported anywhere with a quick `rg` search over the files.

File: fuel-derive/src/attribute.rs
Comment (Voxelot) [reply fuel-derive/src/attribute.rs:68]: According to the spec, aren't the prefixes generally supposed to be a u8? If word size prefixes matches how we serialized things before, that seems like a bug in the spec implementation.

image

Comment (Voxelot) [reply fuel-derive/src/attribute.rs:68]: Word size prefixes makes sense, given this:
image

File: fuel-types/src/canonical.rs
Comment (Voxelot) [inline fuel-types/src/canonical.rs:650]: should we also have a test for regular a non-inner discriminant, ie. like we use on Input?

```rust
#[canonical(discriminant = InputRepr)]
```

File: fuel-derive/manual_impl_tx_serde.rs
Comment (Voxelot) [reply fuel-derive/manual_impl_tx_serde.rs:1]: if it's just for local testing, maybe add a comment explaining what it's for

**Comment by Voxelot** [comment]
What happens when an enum uses `#[canonical(inner_discriminant = Type)]`, but the corresponding structs of each variant don't use `#[canonical(prefix = Type::P1)]`? If we always need to specify the prefix along with discriminant, should we have some kind of check at compile time to ensure it's setup correctly?

File: fuel-types/src/canonical.rs
Comment (Dentosal) [reply fuel-types/src/canonical.rs:650]: Yes. Fortunately it was even easier than that; I simply removed it completely, since it was only used when manually deserializing anyway. 1e14e956

File: fuel-derive/manual_impl_tx_serde.rs
Comment (Dentosal) [reply fuel-derive/manual_impl_tx_serde.rs]: It's not needed anymore. Removed.

File: fuel-derive/Cargo.toml
Comment (Dentosal) [reply fuel-derive/Cargo.toml:1]: Added in b16268cb

File: fuel-tx/src/transaction/types/input/message.rs
Comment (Dentosal) [reply fuel-tx/src/transaction/types/input/message.rs]: 79d57179

File: fuel-tx/src/transaction/types/input/coin.rs
Comment (Dentosal) [reply fuel-tx/src/transaction/types/input/coin.rs]: 79d57179

**Comment by Voxelot** [comment]
Good call on removing discriminant, it was a little confusing how it was suppose to be used compared to inner_descriminant.

**Review by Voxelot — APPROVED** [review]
Wrote some local fuzz tests to check the ability of roundtrip serialization between this and the original encoding, and didn't find any issues. I think we're probably good to ship this.

**Pull request merged**
